examples/chat.rs

```diff
@@ -0,0 +1,125 @@
+use std::collections::HashMap;
+use std::net::SocketAddr;
+use std::sync::{Arc, Mutex};
+
+use futures::{sink::SinkExt, stream::StreamExt};
+
+use tokio::sync::broadcast;
+
+use axum::handler::get;
+use axum::response::Html;
+use axum::routing::RoutingDsl;
+use axum::ws::{ws, Message, WebSocket};
+use axum::{extract, route, AddExtensionLayer};
+
+// Our shared state
+struct AppState {
+    user_map: Mutex<HashMap<String, ()>>,
+    tx: broadcast::Sender<String>,
+}
+
+#[tokio::main]
+async fn main() {
+    let user_map = Mutex::new(HashMap::new());
+    let (tx, _rx) = broadcast::channel(100);
+
+    let app_state = Arc::new(AppState { user_map, tx });
+
+    let app = route("/", get(index))
+        .route("/websocket", ws(websocket))
+        .layer(AddExtensionLayer::new(app_state));
+
+    let addr = SocketAddr::from(([127, 0, 0, 1], 3000));
+
+    hyper::Server::bind(&addr)
+        .serve(app.into_make_service())
+        .await
+        .unwrap();
+}
+
+async fn websocket(stream: WebSocket, state: extract::Extension<Arc<AppState>>) {
+    let state = state.0;
+
+    // By splitting we can send and receive at the same time.
+    let (mut sender, mut receiver) = stream.split();
+
+    // Empty string.
```

Comment:
```suggestion
    // Username gets set in the receive loop, if its valid
```

Probably a bit more descriptive.